New importer package

gimenete opened the pull request:
This PR adds a new package. Maybe there are a few tweaks to do it but I think it's in a good shape to be used. More information on how to use it in the README and in the command's `--help`.

The main goal in this PR was to make a CLI app. I've tried to make the core functionality abstracted enough so it can be used as an API, but I haven't tried that yet. I'll do another PR for further changes regarding that.

Right now the only format supported is CSV, but similarely I've tried to abstract things enough to support more file formats in the future.

This PR supports creating a table if it doesn't exist as well as updating an existing one. It also supports creating new columns in an existing columns if neccessary. It tries to guess the right data type for each column, etc.

SferaDev commented on .changeset/config.json:
```diff
@@ -2,7 +2,7 @@
   "changelog": "@changesets/cli/changelog",
   "commit": false,
-  "linked": [["@xata.io/client", "@xata.io/codegen"]],
+  "linked": [["@xata.io/client", "@xata.io/codegen", "@xata.io/importer"]],
```
Not sure if we want to link all dependency versions. The codegen kind of made sense, but maybe no need to bump all minor/major changes for all of them at the same time.

SferaDev commented on importer/src/cli.ts:
```diff
@@ -0,0 +1,21 @@
+  .option('--format <format>', 'The format of the input source.', 'csv')
+  .option('--create', "Whether the table or columns should be created if they don't exist.", false)
+  .option('--force', 'Whather confirmation should be asked when creating the table or columns.', false)
+  .option('--branch', 'The branch to import data to.', 'main')
```
```suggestion
  .option('--branch <branch>', 'The branch to import data to.', 'main')
```

SferaDev commented on importer/src/cli.ts:
```diff
@@ -0,0 +1,21 @@
+  .option('--types <types>', 'List of column names.')
+  .option('--noheader', 'Specify that the CSV file has no header.')
+  .option('--format <format>', 'The format of the input source.', 'csv')
+  .option('--create', "Whether the table or columns should be created if they don't exist.", false)
```
Not sure I get the ``--create`` flag. If I don't pass it, it errors me out. If I do pass it still asks me for confirmation.

In case you don't pass it I would ask interactively and have the flag to not ask.

<img width="758" alt="Screenshot 2022-05-04 at 08 20 19" src="https://user-images.githubusercontent.com/2181866/166631559-8b22a439-3714-474a-9bb1-ebf575bd66f5.png">

SferaDev replied:
Yeah, although I would make it always ask and maybe have a interactive/non-interactive flag that disables inquirer everywhere and fails if the flags for ``create`` or ``force`` are not provided. WDYT?

SferaDev commented on importer/src/cli.ts:
```diff
@@ -0,0 +1,21 @@
+  .option('--table <table>', 'The table in which to import the data.')
+  .option('--columns <columns>', 'List of column names.')
+  .option('--types <types>', 'List of column names.')
+  .option('--noheader', 'Specify that the CSV file has no header.')
```
I would have expected with this flag to use the existing table columns, but I guess order cannot be guaranteed. Could we make it interactive somehow? [just a comment, but I understand this as a future improvement]

SferaDev reviewed (commented):
I would like to add new rows to an already existing table/schema. However it's not properly comparing the types from the online schema.

<img width="758" alt="Screenshot 2022-05-04 at 08 44 48" src="https://user-images.githubusercontent.com/2181866/166633830-09e03660-9989-429e-95d9-bcaffb30428b.png">

<img width="271" alt="Screenshot 2022-05-04 at 08 45 03" src="https://user-images.githubusercontent.com/2181866/166633848-7d8f329e-5e5f-4f3b-a16c-421a9e7b6cef.png">

<img width="429" alt="Screenshot 2022-05-04 at 08 29 07" src="https://user-images.githubusercontent.com/2181866/166632289-f0a1b723-cbd7-411f-b006-15f95589a2a5.png">

SferaDev commented:
Since we rely on the client we also have the createMany error handling issue

<img width="758" alt="Screenshot 2022-05-04 at 08 31 36" src="https://user-images.githubusercontent.com/2181866/166632546-ed8bb4e2-5647-445a-b18f-c7a5a8397acb.png">

![giphy (1)](https://user-images.githubusercontent.com/2181866/166632657-aec0377b-12e1-446e-9f95-16dfb75e8fa6.gif)

SferaDev commented on importer/package.json:
```diff
@@ -0,0 +1,44 @@
+    "camelcase": "^6.3.0",
+    "commander": "^9.2.0",
+    "cross-fetch": "^3.1.5",
+    "csvtojson": "^2.0.10",
```
Not 100% sure about this dependency, seems not actively maintained

<img width="792" alt="Screenshot 2022-05-04 at 08 37 34" src="https://user-images.githubusercontent.com/2181866/166633074-e6c7fc66-a640-4f30-be45-5471d006989f.png">

SferaDev commented on importer/src/cli.ts:
```diff
@@ -0,0 +1,21 @@
+  .name('xata-importer')
+  .description('Importer utility for xata.io')
+  .version(IMPORTER_VERSION)
+  .argument('<file>', 'File containing the data to import. Pass "-" to pipe content instead.')
```
I would add another option for ``delimiter`` and probably attempt to make it "smart".

Although CSV stands for comma-separated-values, the delimiter could be anything. Most popular are semicolons, tabs, spaces, etc...

gimenete replied:
Good idea.

Question about one use case for that. How could we specify tab delimiters with flags in the CLI? 🤔

SferaDev replied:
What about "\t"

SferaDev commented on importer/src/csv.ts:
```diff
@@ -0,0 +1,56 @@
+}
+
+export async function parseFile(file: string, options: ParseOptions) {
+  return process(initConverter(options).fromFile(file), options);
```
Do we handle different encodings? I wouldn't assume UTF-8 by default, and maybe make it configurable

gimenete replied:
Good suggestion. I'm not sure if the csv parser already takes care of that somehow (checking if there's a BOM for UTF), but I guess for most cases if the encoding is a different one, we'd need a way to specify it.

SferaDev reviewed (commented):
@gimenete Works really good, I love the type inference 💯 

Added a lot of comments, most of them can be taken care after merging. I'll leave you decide how you want to prioritise them. Ping me if you need an approval.

gimenete commented:
> Since we rely on the client we also have the createMany error handling issue

Yup. I'll upgrade it once we release a new version of the client.

> I would like to add new rows to an already existing table/schema. However it's not properly comparing the types from the online schema. 

I'll look into that.

Thanks for your review!! 🙏

gimenete commented:
@SferaDev the problem with the email/string thing should be fixed. I was not implementing the easiest case for `castType`: when the types are equal 🤦‍♂️ 

I've addressed some feedback and I think the rest can be implemented or explored in other PRs.

SferaDev reviewed (approved):
Tested again, now it works properly.

I would create several follow-up issues:

- Add delimiter option (semicolons, tabs, spaces)
- Review different encodings (latin1)
- Add a non-interactive flag, make all commands to inquiry instead of erroring out, if flags are provided ignore the inquiry question (related to https://github.com/xataio/client-ts/pull/127/files#r864487665)
- Add a way to call interactively as an npm package (API)

gimenete commented:
Thanks @SferaDev for the review and the feedback! I've created https://github.com/xataio/client-ts/issues/142
